### ISSUE TITLE
Wine 6.1 staging (32bit) : BK Intro VST crashes every host with SIGABRT

### ISSUE THREAD

**Issue by artscoop** [robbert-vdh/yabridge#81]
Thanks for giving yabridge a shot! If you're having issues with plugins not working at all or scans timing out then make sure to check out the [troubleshooting common issues](https://github.com/robbert-vdh/yabridge#troubleshooting-common-issues) section of the readme.

**Problem description**
When I register the VSTi plugin BK Intro (32 bit), it is recognized by the host as a valid VSTi. However, trying to open the plugin crashes the host application with a SIGABRT (happened in Ardour, Carla and Renoise).
You just need a 32 or 64 bit Wine prefix and the following plugin, and Wine staging 6.1
[BkSynthlabs Intro.tar.gz](https://github.com/robbert-vdh/yabridge/files/5975893/BkSynthlabs.Intro.tar.gz)

**What did you expect to happen?**
The plugin GUI opens, or, for Carla, the plugin parameter list loads.

**What actually happened?**
Immediate crash of the host, with a SIGABRT signal.

**System information**
- Plugin: Boris Kovalev INTRO
- Plugin type: VST2
- Host: Ardour, Renoise, Carla
- Operating system: Manjaro
- Wine version: Wine Staging 6.1
- Installation method: copies, manual
- yabridge version: 2.2.1, commit 3fbffa5
- yabridgectl version: Not used

**Debug log**
[yabridge.log](https://github.com/robbert-vdh/yabridge/files/5975898/yabridge.log)

Terminal output:
```
** (ardour-6.5.0:25828): ERROR **: 16:20:07.420: 
unhandled exception (type std::exception) in signal handler:
what: read: End of file

fish : Tâche , 'env YABRIDGE_DEBUG_FILE=/tmp/ya…' terminée par le signal SIGTRAP (Déroutement de suivi/point d'arrêt)
```

**Comment by robbert-vdh** [comment]
The plugin (downloaded from https://vst4free.com/plugin/324/) works fine here with the latest master branch (2.2.1-777-g907925d) and with both Wine Staging 6.1 from the Arch/Manjaro repos and with a wine-tkg version based on Wine 6.1.r7 in a new clean Wine prefix. Since the plugin seems to block on handling Win32 messages (and that `effMainsChanged()` call at the end has to wait for those to be finished because of a bug in EZdrummer), this is likely caused by some library override. I'd try creating a new a clean Wine prefix and loading the plugin from there.

**Comment by artscoop** [comment]
Damn Robert you're quick!
I'm going to test the plugin again tonight in a fresh Wine prefix.
I'm also going to give you the list of plugins overridden in my current prefix

**Comment by artscoop** [comment]
Okay, so I've created a new Wine prefix (which is default), redownloaded BK Intro from vst4free and recreated the .so file.
The same error happened again, I'm pretty confused as why it works on your machine.
I am even using a pretty fresh commit of yabridge from yesterday (44cb2ff).

One difference between your machine and mine may be how yabridge is built. I'm using `yabridge-git` from AUR (it always builds from the latest commit available on master)
Here is the pkgbuild info, maybe there are some options causing issues:

```
# Maintainer: Robbert van der Helm <mail@robbertvanderhelm.nl>

pkgname=yabridge-git
_pkgname=yabridge
pkgver=2.2.1.r776.g4f8eaaa
pkgrel=1
pkgdesc="Yet Another VST bridge, run Windows VST2 plugins under Linux"
arch=('x86_64')
url="https://github.com/robbert-vdh/yabridge"
license=('GPL3')
depends=('wine' 'boost'  'libxcb' 'lib32-boost-libs>=1.72.0' 'lib32-libxcb')
makedepends=('meson' 'ninja')
optdepends=('yabridgectl: utility for setting up and managing yabridge')
provides=('yabridge')
conflicts=('yabridge')
options=('!strip')
install=yabridge.install
source=('git+https://github.com/robbert-vdh/yabridge')
sha256sums=('SKIP')

pkgver() {
  cd "$_pkgname"
  git describe --long | sed 's/\([^-]*-g\)/r\1/;s/-/./g'
}

build() {
  cd "$_pkgname"

  # Meson won't apply any new options or update wraps if they already exist, so
  # if we're building from a dirty src/ directory we'll just nuke any cached
  # files
  if [[ -d build ]]; then
    rm -rf build

    # Can't use `git clean` here since the entire src/ directory will be
    # gitignored
    find ./subprojects -mindepth 1 -maxdepth 1 -type d -exec rm -rf '{}' ';'
  fi

  # If you don't want to build lib32-boost-libs and you don't need the 32-bit
  # bitbridge, then you can leave out the dependency for it and set the
  # `use-bitbridge` option to false.
  # You can also add `-Dwith-vst3=false` to disable building with VST3 support.
  # If building takes up too much RAM then you can decrease the unity size or
  # completely disable unity builds.
  meson setup \
    --cross-file cross-wine.conf \
    --buildtype=release \
    --unity=on \
    --unity-size=1000 \
    -Dwith-bitbridge=true \
    build

  # The unity build takes can take up to 2 GB of RAM per target, so if the
  # system does not have enough RAM to build everything at once we'll limit the
  # number of concurrent jobs
  total_memory=$(free --gibi --si | awk '/^Mem:/ { print $2 }')
  if [ "$total_memory" -le 4 ]; then
    echo -e "\n$total_memory gigabytes of RAM detected, limiting the number of build jobs to 1\n"
    ninja -C build -j1
  elif [ "$total_memory" -le 8 ]; then
    echo -e "\n$total_memory gigabytes of RAM detected, limiting the number of build jobs to 3\n"
    ninja -C build -j3
  else
    ninja -C build
  fi
}
```

**Comment by robbert-vdh** [comment]
`read: End of file` would indicate that the Wine plugin host process got terminated. There's no Wine error in the log that would point to a cause though. Perhaps try installing the optional `lib32-` dependencies from `pacman -Qi wine` if you're missing any of those? I have all of them (and the non-multilib equivalents) installed here.

**Comment by artscoop** [comment]
Got it, I'm doing it right now and reporting soon after. Thanks!

**Comment by artscoop** [comment]
So, I installed every lib32- optional dependency (there were 3 I did not have installed)
Rebuilt yabridge from AUR
Recreated the .so files fr all VST plugins
Cleared Carla and rescanned VSTs
But still the same crash with SIGABRT.

I'll have to try this on another machine, probably tomorrow 🤔

**Comment by robbert-vdh** [comment]
Do any other 32-bit VST2 plugins work? Or any 64-bit plugins for that matter? For instance, [FerricTDS](https://varietyofsound.wordpress.com/2009/10/29/ferrictds-released-today-withi-the-kvr-dc09/. You'll have to extract the .zip archive with the command line `unzip` tool or you might have to manually rename the files because the archive is malformed and uses backslashes. The read error would indicate that the Wine process got terminated, so you'd need to figure out why that's happening. There's sadly not really a convenient way to do that I could recommend to you.

**Comment by artscoop** [comment]
Most other 32-bit VSTs in my collection (about 250) seem to work with Yabridge.
Another plugin, Linplug SaxLab, provokes a crash of yabridge-server-32.exe but it's non-fatal and happens also in Wine apps (OpenMPT) so Yabridge is having exactly what Windows has.
I've tried FerricTDS and the plugin loads perfectly fine (after renaming the output files)

I'll try to find why such things can happen, if I find something relevant I'll put it here

**Comment by artscoop** [comment]
Re,
I've tried some other things recently which bring me to think that there is inherently suspect in Wine.

Completely unrelated to yabridge, at some time in the past, for no good, reason, I could not get a DirectX9 app to run in Wine, when months prior, it ran absolutely fine. I was using Ubuntu at the time. But I think it's similar to what's happening now.

Something weird in the same vein appeared with LinVst (another bridge). I had a recently installed Manjaro with a set of maybe 30 VSTs, 64-bt mostly, that would work properly in Renoise Linux. Then, months after and a few updates of many packages later, LinVst started behaving where opening 2 VSTs in the same bridge and closing one VST would freeze the linvst server process. May be due to a change in LinVst.

Then, there is yabridge. As you know with this issue, Yabridge works better than Linvst on my two Manjaro systems, one with 250+ 32-bit VSTs and the other with around 25 64-bit VSTs. Still, as you know, at least one completely innocuous VST gets processes killed with a SIGTRAP or SIGABRT.
I tried using WINEDEBUG to trace what was going on, and I could just see that the plugin could load a few of its Synthedit files, then crash without trace before loading the Synthedit wavetable file.
Then, I made myself a new Virtualbox VM, installed a new Manjaro, applied updates, installed Pipewire, installed your yabridge AUR package (which installed Wine but not Wine staging), then installed Carla. I opened Carla to test the BK Synthlab Intro plugin, and... no crash. The plugin works as well as on your machine.
On my host machine (not in my virtual machine), I made a all new Wine default prefix, tried using the Intro VST... Crash.

So, the difference between Intro working and Intro not working might be one of those things, with some thing probably unrelated:
- Video driver used (Intel+Nouveau/Nvidia where it crashes)
- Maybe one system config file (limits.conf or something else)
- Wine vs Wine staging (Wine staging where it crashes)
- My Manjaro system where it crashes has an old /home directory (older than the current Manjaro install)
- I might have one package installed on my machine that isn't on my VM (or vice-versa) that results in a SIGABRT (maybe something related to libc or libboost)

**Comment by robbert-vdh** [comment]
Do you have more of those plugins that have this behaviour? Perhaps I can reproduce it with other plugins. If it's in any way related to yabridge then I might be able to trace it down using gdb, but attaching GDB to a Winelib doesn't really work (there are a few ways to do it, all of them have different caveats), so I can't really ask you to try it because it involves a bit of manual setup.

**Comment by robbert-vdh** [comment]
Is this still an issue with current Wine and yabridge 3.1.0?

**Comment by artscoop** [comment]
Hi Robbert, I had trouble making any VST work with Wine 6.5 (I think it was broken and Wine 6.6 fixed that), but I'll check again today if I have some time. I see that Yabridge 3.1 was released today in Arch, I tested it one hour ago with some plugins that were working before and it worked as expected.
What I still need to test is BK Intro, I'll check and report.

**Comment by daniel-j** [comment]
I get this error with MT Power Drum Kit 2 (Linux version is a Windows dll) 64-bit version. It works fine on my Fedora machine (using wine 6.4 from Lutris since wine 6.3 in Fedora packages is broken with yabridge), but the same setup doesn't work on my Arch Linux machine, resulting in the `read: End of file` error. Removed `~/.wine` directory prior. However, I tried another Windows VST which worked fine (Youlean Loudness Meter).

I am using Reaper. The wine versions in Arch Linux (6.6-1 at the moment) shows the same error as Lutris 6.4 wine on Arch.

I sync like this:
`YABRIDGE_DEBUG_FILE=/tmp/yabridge.log PATH=$PATH:/home/djazz/.local/share/lutris/runners/wine/lutris-6.4-x86_64/bin yabridgectl sync`
And launch Reaper like this:
`YABRIDGE_DEBUG_FILE=/tmp/yabridge.log PATH=$PATH:/home/djazz/.local/share/lutris/runners/wine/lutris-6.4-x86_64/bin reaper`

Relevant log from Arch Linux (using wine lutris-6.4) when loading the VSTi:
```
20:28:53 [MT-PowerDrumKit-58DINdfi] Initializing yabridge version 3.1.0
20:28:53 [MT-PowerDrumKit-58DINdfi] host:          '/home/djazz/.local/share/yabridge/yabridge-host.exe'
20:28:53 [MT-PowerDrumKit-58DINdfi] plugin:        '/home/djazz/.vst/MT/MT-PowerDrumKit.dll'
20:28:53 [MT-PowerDrumKit-58DINdfi] plugin type:   'VST2'
20:28:53 [MT-PowerDrumKit-58DINdfi] realtime:      'yes'
20:28:53 [MT-PowerDrumKit-58DINdfi] sockets:       '/run/user/1000/yabridge-MT-PowerDrumKit-58DINdfi'
20:28:53 [MT-PowerDrumKit-58DINdfi] wine prefix:   '<default>'
20:28:53 [MT-PowerDrumKit-58DINdfi] wine version:  '6.4.r0.g7ec998e1 ( TkG Staging Esync Fsync )'
20:28:53 [MT-PowerDrumKit-58DINdfi] 
20:28:53 [MT-PowerDrumKit-58DINdfi] config from:   '<defaults>'
20:28:53 [MT-PowerDrumKit-58DINdfi] hosting mode:  'individually, 64-bit'
20:28:53 [MT-PowerDrumKit-58DINdfi] other options: '<none>'
20:28:53 [MT-PowerDrumKit-58DINdfi] 
20:28:53 [MT-PowerDrumKit-58DINdfi] Enabled features:
20:28:53 [MT-PowerDrumKit-58DINdfi] - bitbridge support
20:28:53 [MT-PowerDrumKit-58DINdfi] - VST3 support
20:28:53 [MT-PowerDrumKit-58DINdfi] 
20:28:53 [MT-PowerDrumKit-58DINdfi] [Wine STDERR] wineserver: using server-side synchronization.
20:28:53 [MT-PowerDrumKit-58DINdfi] [Wine STDERR] wine: Using setpriority to control niceness in the [-10,10] range
20:28:53 [MT-PowerDrumKit-58DINdfi] [Wine STDERR] 002c:fixme:winediag:LdrInitializeThunk Wine TkG (staging) 6.4 is a testing version containing experimental patches.
20:28:53 [MT-PowerDrumKit-58DINdfi] [Wine STDERR] 002c:fixme:winediag:LdrInitializeThunk Please don't report bugs about it on winehq.org and use https://github.com/Frogging-Family/wine-tkg-git/issues instead.
20:28:54 [MT-PowerDrumKit-58DINdfi] [Wine STDOUT] Initializing yabridge host version 3.1.0
20:28:54 [MT-PowerDrumKit-58DINdfi] [Wine STDOUT] Preparing to load VST2 plugin at '/home/djazz/.vst/MT/MT-PowerDrumKit.dll'
20:28:58 [MT-PowerDrumKit-58DINdfi] [Wine STDERR] 0090:err:rpc:I_RpcReceive we got fault packet with status 0x1c010003
20:28:58 Error during initialization:
20:28:58 read: End of file
```

**Comment by robbert-vdh** [comment]
@daniel-j Could you open another ticket for this? And as a question for in that ticket, do you also get that `0090:err:rpc:I_RpcReceive we got fault packet with status 0x1c010003` error with Youlean Loudness Meter? I don't have any issues with MT Pwer Drumkit 2.0.9, also with wine-tkg-6.4.r0.g7ec998e1, on both my usual and on a clean Wine prefix. I also don't get that `err:rpc:I_RpcReceive` error, so it must be something specific to your setup.

**Comment by daniel-j** [comment]
> @daniel-j do you also get that `0090:err:rpc:I_RpcReceive we got fault packet with status 0x1c010003` error with Youlean Loudness Meter?
Nope, no wine warnings there
```
20:49:21 [Youlean Loudness Meter 2-ILRWahiq] [Wine STDOUT] Preparing to load VST2 plugin at '/home/djazz/.vst/Youlean/Youlean Loudness Meter 2.dll'
20:49:22 [Youlean Loudness Meter 2-ILRWahiq] [Wine STDOUT] Finished initializing '/home/djazz/.vst/Youlean/Youlean Loudness Meter 2.dll'
20:49:22 [Youlean Loudness Meter 2-ILRWahiq] 
20:49:22 [Youlean Loudness Meter 2-ILRWahiq]    The host has requested libSwell GUI support, which is
20:49:22 [Youlean Loudness Meter 2-ILRWahiq]    not supported when using Wine.
20:49:22 [Youlean Loudness Meter 2-ILRWahiq]    You can safely ignore this message; this is normal
20:49:22 [Youlean Loudness Meter 2-ILRWahiq]    when using REAPER.
20:49:22 [Youlean Loudness Meter 2-ILRWahiq] 
```
I'll open a new issue.

**Comment by daniel-j** [comment]
Re: This issue, I tried the BK Intro VST and REAPER crashes with the segfault. So not the same as my comment above.

**Comment by robbert-vdh** [comment]
The plugins still working fine here:

![afbeelding](https://user-images.githubusercontent.com/748520/115128559-e9437380-9fde-11eb-847d-0a4228f76da1.png)

You might be missing some multilib (`lib32-`) optional Wine dependency. There's probably some relevant Wine output just before the segfault.

**Comment by robbert-vdh** [comment]
So @artscoop, were you by any chance using PipeWire on your main machine? If so, then that was with 100% certainty the issue here. PipeWire doesn't use Linux's regular realtime scheduling system but it uses rtkit instead, and both PipeWire and rtkit try to place limits on the amount of CPU time a realtime process may use. This causes yabridge's process to get terminated while initializing the plugin (which can take longer than the allowed couple of milliseconds). You should technically be able to increase this limit, but you'd need to configure both PipeWire and rtkit for that to work. I'd recommend just using regular JACK2 for the time being, especially since PipeWire's performance and latency also aren't nearly as good as JACK or pure ALSA yet, and it's still missing a few JACK features.

**Comment by artscoop** [comment]
Oh damn, excellent catch. I actually use Pipewire on all my machines.
I've also been using Carla to test VSTs in a Virtualbox, and I could not understand how I could never load any VST anymore (either freeze or immediate "Could not load plugin"). I've just tested removing pipewire-pulse, pipewire-jack, and reinstalled pulseaudio and jack2-dbus + Cadence, and even BK Intro could load in Carla.

Now I have to try this on my main machine but your insight is outstanding. How did you get this information?
(I might also try configuring Pipewire and RTkit for science)

**Comment by robbert-vdh** [comment]
#92 and #77 also had similar issues with crashing plugins that I was unable to reproduce, and the only thing all of those reports had in common was that they all used PipeWire. In all of those cases it seemed like the Wine process just shut down, but without a Wine error or segfault. It turned out that launching `yabridge-host.exe` from a DAW running under PipeWire would get `SIGKILL`'d, but launching that same `yabridge-host.exe` process manually from another terminal would work fine. I wasn't able to figure out the cause of the `SIGKILL` using the usual methods (dmesg/journalctl, auditd, gdb and strace), but I had seen processes get killed before because of CPU time limits in realtime scheduling. Only difference is that normally you'd get a `SIGXCPU`, and not a `SIGKILL`. So I tried disabling realtime scheduling in yabridge's Wine host applications, and that stopped the `SIGKILL` from happening. After asking around on `#pipewire` PipeWire's lead dev immediately pointed me to the realtime CPU time limit set by both PipeWire and rtkit (through `RLIMIT_RTTIME`). Normally that's set to unlimited, but using PipeWire reduces that to a value somewhere in the tens or hundreds of milliseconds. When a plugin takes longer than that to initialize on an audio thread, then the host will get terminated. This of course also affects native plugins.

So you would need to configure both PipeWire and rtkit to just leave that limit at unlimited/-1 (or crank it up to a few orders of magnitudes). Or you can just not use PipeWire, that's probably the easiest solution for now. :sweat_smile: At some point in the future they're going to use the regular realtime scheduling system (the thing you would need to give your user permissions for) and only fall back to rtkit if that's unavailable, but I don't know when that will be.

**Comment by artscoop** [comment]
Hi Robbert,
So, indeed, I tried BK Intro VST on my main system (the one that failed miserably), and with yabridge and jack2, it worked! That's a relief, even though I like the simplicity of Pipewire.

Anecdotally, the Vital VST and standalone app, which is native for Linux, used to crash everytime I tried to install the default bundle files. Now that I don't use Pipewire-jack, it does not crash anymore, yet, the standalone app is supposed to use ALSA and not Jack.
I'm curious to see how and when the Pipewire developers will come up with a solution not needing rtkit.

Thanks very much for your help and for your piece of software!
